desktop/core/src/desktop/lib/fs/gc/gs.py

```diff
@@ -0,0 +1,454 @@
+      except GSFileSystemException as e:
+        raise e
+      except GSResponseError as e:
+        if 'Forbidden' in str(e) or (hasattr(e, 'status') and e.status == 403):
+          raise GSListAllBucketsException(
+            _('You do not have permissions to list all buckets. Please specify a bucket name you have access to.'))
+        else:
+          raise GSFileSystemException(_('Failed to retrieve buckets: %s') % e.reason)
+      except Exception as e:
+        raise GSFileSystemException(('Failed to retrieve buckets: %s') % e)
+
+    bucket_name, prefix = parse_uri(path)[:2]
+    bucket = self._get_bucket(bucket_name)
+    prefix = self._append_separator(prefix)
+
+    res = []
+    for item in bucket.list(prefix=prefix, delimiter='/', headers=self.header_values):
+      if isinstance(item, Prefix):
+        res.append(GSStat.from_key(Key(item.bucket, item.name), is_dir=True, fs=self.fs))
+      else:
+        if item.name == prefix:
+          continue
+        res.append(self._stats_key(item, self.fs))
+
+    return res
+
+  @translate_gs_error
+  def listdir(self, path, glob=None):
+    return [parse_uri(x.path)[2] for x in self.listdir_stats(path, glob)]
+
+  @translate_gs_error
+  @auth_error_handler
+  def rmtree(self, path, skipTrash=True):
+    """Remove keys from GS filesystem.
+
+    Args:
+      path (str): The GS key path of the file or directory to remove.
+      skipTrash (bool): Whether to skip the trash when deleting.
+
+    Raises:
+      NotImplementedError: Since moving to trash is not implemented.
+      GSFileSystemException: If the removal operation fails.
+    """
+    if not skipTrash:
+      raise NotImplementedError(_('Moving to trash is not implemented for GS'))
+
+    bucket_name, key_name = parse_uri(path)[:2]
+    if bucket_name and not key_name:
+      self._delete_bucket(bucket_name)
+    else:
+      if self.isdir(path):
+        # Really need to make sure we end with a '/' for directory and it reflects in key_name
+        path = self._append_separator(path)
+        _, key_name = parse_uri(path)[:2]
+
+      key = self._get_key(path)
+      if key:
+        dir_keys = []
+        if self.isdir(path):
+          dir_keys = key.bucket.list(prefix=key_name)
+
+        if not dir_keys:
+          # Avoid Raz bulk delete issue
+          deleted_key = key.delete()
+          if deleted_key.exists():
+            raise GSFileSystemException('Could not delete key %s' % deleted_key)
+        else:
+          # key.bucket.delete_keys() call is not supported from GS side
+          # So, try deleting the all keys with directory prefix one by one
+
+          # TODO: Check on the UI side if key count is greater than max limit and show nice notification.
+          deleted_dir_key_count = 0
+          for key in list(dir_keys):
+            if deleted_dir_key_count > GS_BULK_DELETE_DIR_KEYS_MAX_LIMIT.get():
+              break
```

Comment:
We should maybe send an error back to the UI stating the limit crossed and that we deleted only x number of files. They can either retry or follow the platform's recommendations to delete large number of files.

Reply:
Yes I added a TODO above to check the limit before even calling the operation on UI side and send a notification or a dialogue box. That is some UI change for which I'll need help so we can take that up in a separate PR.